File: backend/utils/research_error_handler.py
"""
Research Error Handling Utilities
Provides robust error handling and retry logic for customer research operations.
"""

import asyncio
import logging
import time
from typing import Any, Callable, Dict, Optional, Tuple, TypeVar, Union
from functools import wraps
from backend.config.research_config import RESEARCH_CONFIG, ERROR_CONFIG

logger = logging.getLogger(__name__)

T = TypeVar('T')

class ResearchError(Exception):
    """Base exception for research-related errors"""
    def __init__(self, message: str, error_code: str = "research_error", retry_after: Optional[int] = None):
        self.message = message
        self.error_code = error_code
        self.retry_after = retry_after
        super().__init__(self.message)

class APITimeoutError(ResearchError):
    """Exception for API timeout errors"""
    def __init__(self, message: str = None):
        super().__init__(
            message or ERROR_CONFIG.ERROR_MESSAGES["api_timeout"],
            "api_timeout",
            retry_after=5
        )

class APIError(ResearchError):
    """Exception for general API errors"""
    def __init__(self, message: str = None):
        super().__init__(
            message or ERROR_CONFIG.ERROR_MESSAGES["api_error"],
            "api_error",
            retry_after=3
        )

class RateLimitError(ResearchError):
    """Exception for rate limiting errors"""
    def __init__(self, message: str = None, retry_after: int = 60):
        super().__init__(
            message or ERROR_CONFIG.ERROR_MESSAGES["rate_limit"],
            "rate_limit",
            retry_after=retry_after
        )

class ServiceUnavailableError(ResearchError):
    """Exception for service unavailable errors"""
    def __init__(self, message: str = None):
        super().__init__(
            message or ERROR_CONFIG.ERROR_MESSAGES["service_unavailable"],
            "service_unavailable",
            retry_after=30
        )

class ErrorHandler:
    """Centralized error handling for research operations"""
    
    @staticmethod
    def get_fallback_response(stage: str = "general") -> str:
        """Get a fallback response for the given stage"""
        return ERROR_CONFIG.FALLBACK_RESPONSES.get(stage, ERROR_CONFIG.FALLBACK_RESPONSES["general"])
    
    @staticmethod
    def handle_llm_error(error: Exception, context: Dict[str, Any] = None) -> Tuple[str, str]:
        """
        Handle LLM service errors and return appropriate response
        
        Args:
            error: The exception that occurred
            context: Additional context about the error
            
        Returns:
            Tuple of (response_message, error_code)
        """
        context = context or {}
        stage = context.get('stage', 'general')
        
        if isinstance(error, asyncio.TimeoutError):
            logger.warning(f"LLM timeout error: {error}")
            return ERROR_CONFIG.ERROR_MESSAGES["api_timeout"], "api_timeout"
        
        elif "rate limit" in str(error).lower():
            logger.warning(f"LLM rate limit error: {error}")
            return ERROR_CONFIG.ERROR_MESSAGES["rate_limit"], "rate_limit"
        
        elif "service unavailable" in str(error).lower() or "503" in str(error):
            logger.error(f"LLM service unavailable: {error}")
            return ERROR_CONFIG.ERROR_MESSAGES["service_unavailable"], "service_unavailable"
        
        else:
            logger.error(f"LLM general error: {error}")
            # Return a fallback response based on conversation stage
            fallback = ErrorHandler.get_fallback_response(stage)
            return fallback, "api_error"
    
    @staticmethod
    def handle_validation_error(error: Exception) -> Tuple[str, str]:
        """Handle validation errors"""
        logger.warning(f"Validation error: {error}")
        return ERROR_CONFIG.ERROR_MESSAGES["validation_error"], "validation_error"
    
    @staticmethod
    def handle_session_error(error: Exception) -> Tuple[str, str]:
        """Handle session-related errors"""
        logger.warning(f"Session error: {error}")
        return ERROR_CONFIG.ERROR_MESSAGES["session_expired"], "session_expired"

def with_retry(
    max_retries: int = None,
    delay: float = None,
    exponential_backoff: bool = None,
    exceptions: Tuple[Exception, ...] = (Exception,)
):
    """
    Decorator to add retry logic to functions
    
    Args:
        max_retries: Maximum number of retries (defaults to config)
        delay: Initial delay between retries (defaults to config)
        exponential_backoff: Whether to use exponential backoff (defaults to config)
        exceptions: Tuple of exceptions to retry on
    """
    if max_retries is None:
        max_retries = RESEARCH_CONFIG.MAX_RETRIES
    if delay is None:
        delay = RESEARCH_CONFIG.RETRY_DELAY_SECONDS
    if exponential_backoff is None:
        exponential_backoff = ERROR_CONFIG.EXPONENTIAL_BACKOFF
    
    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        @wraps(func)
        async def async_wrapper(*args, **kwargs) -> T:
            last_exception = None
            current_delay = delay
            
            for attempt in range(max_retries + 1):
                try:
                    return await func(*args, **kwargs)
                except exceptions as e:
                    last_exception = e
                    
                    if attempt == max_retries:
                        logger.error(f"Function {func.__name__} failed after {max_retries} retries: {e}")
                        raise e
                    
                    logger.warning(f"Function {func.__name__} failed (attempt {attempt + 1}/{max_retries + 1}): {e}")
                    
                    if current_delay > 0:
                        await asyncio.sleep(current_delay)
                    
                    if exponential_backoff:
                        current_delay = min(current_delay * 2, ERROR_CONFIG.MAX_BACKOFF_SECONDS)
            
            # This should never be reached, but just in case
            raise last_exception
        
        @wraps(func)
        def sync_wrapper(*args, **kwargs) -> T:
            last_exception = None
            current_delay = delay
            
            for attempt in range(max_retries + 1):
                try:
                    return func(*args, **kwargs)
                except exceptions as e:
                    last_exception = e
                    
                    if attempt == max_retries:
                        logger.error(f"Function {func.__name__} failed after {max_retries} retries: {e}")
                        raise e
                    
                    logger.warning(f"Function {func.__name__} failed (attempt {attempt + 1}/{max_retries + 1}): {e}")
                    
                    if current_delay > 0:
                        time.sleep(current_delay)
                    
                    if exponential_backoff:
                        current_delay = min(current_delay * 2, ERROR_CONFIG.MAX_BACKOFF_SECONDS)
            
            # This should never be reached, but just in case
            raise last_exception
        
        # Return appropriate wrapper based on whether function is async
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        else:
            return sync_wrapper
    
    return decorator

def with_timeout(timeout_seconds: int = None):
    """
    Decorator to add timeout to async functions
    
    Args:
        timeout_seconds: Timeout in seconds (defaults to config)
    """
    if timeout_seconds is None:
        timeout_seconds = RESEARCH_CONFIG.REQUEST_TIMEOUT_SECONDS
    
    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        @wraps(func)
        async def wrapper(*args, **kwargs) -> T:
            try:
                return await asyncio.wait_for(func(*args, **kwargs), timeout=timeout_seconds)
            except asyncio.TimeoutError:
                logger.error(f"Function {func.__name__} timed out after {timeout_seconds} seconds")
                raise APITimeoutError()
        
        return wrapper
    
    return decorator

def safe_execute(func: Callable[..., T], *args, fallback_value: T = None, **kwargs) -> T:
    """
    Safely execute a function with error handling
    
    Args:
        func: Function to execute
        *args: Arguments for the function
        fallback_value: Value to return if function fails
        **kwargs: Keyword arguments for the function
        
    Returns:
        Function result or fallback value
    """
    try:
        return func(*args, **kwargs)
    except Exception as e:
        logger.error(f"Safe execution failed for {func.__name__}: {e}")
        return fallback_value

File: frontend/lib/utils/research-error-handler.ts
/**
 * Research Error Handling Utilities
 * Provides robust error handling and retry logic for frontend research operations.
 */

import { RESEARCH_CONFIG } from '@/lib/config/research-config';

export class ResearchError extends Error {
  constructor(
    message: string,
    public errorCode: string = 'research_error',
    public retryAfter?: number
  ) {
    super(message);
    this.name = 'ResearchError';
  }
}

export class NetworkError extends ResearchError {
  constructor(message?: string) {
    super(
      message || RESEARCH_CONFIG.errorMessages.networkError,
      'network_error',
      5
    );
  }
}

export class ValidationError extends ResearchError {
  constructor(message?: string) {
    super(
      message || RESEARCH_CONFIG.errorMessages.validationError,
      'validation_error'
    );
  }
}

export class RateLimitError extends ResearchError {
  constructor(message?: string, retryAfter: number = 60) {
    super(
      message || RESEARCH_CONFIG.errorMessages.rateLimited,
      'rate_limit_error',
      retryAfter
    );
  }
}

export class ServiceUnavailableError extends ResearchError {
  constructor(message?: string) {
    super(
      message || RESEARCH_CONFIG.errorMessages.serviceUnavailable,
      'service_unavailable',
      30
    );
  }
}

export class SessionExpiredError extends ResearchError {
  constructor(message?: string) {
    super(
      message || RESEARCH_CONFIG.errorMessages.sessionExpired,
      'session_expired'
    );
  }
}

export class ErrorHandler {
  /**
   * Handle API errors and return appropriate user-friendly messages
   */
  static handleApiError(error: any): { message: string; errorCode: string; retryAfter?: number } {
    console.error('API Error:', error);

    // Network errors
    if (!navigator.onLine) {
      return {
        message: 'You appear to be offline. Please check your internet connection.',
        errorCode: 'offline'
      };
    }

    // CSP errors
    if (error instanceof TypeError && error.message.includes('Content Security Policy')) {
      return {
        message: 'Connection blocked by security policy. Please ensure the backend is running on the correct port and try refreshing the page.',
        errorCode: 'csp_error',
        retryAfter: 0
      };
    }

    // Fetch errors
    if (error instanceof TypeError && error.message.includes('fetch')) {
      return {
        message: RESEARCH_CONFIG.errorMessages.networkError,
        errorCode: 'network_error',
        retryAfter: 5
      };
    }

    // HTTP errors
    if (error.status) {
      switch (error.status) {
        case 400:
          return {
            message: RESEARCH_CONFIG.errorMessages.validationError,
            errorCode: 'validation_error'
          };
        case 429:
          return {
            message: RESEARCH_CONFIG.errorMessages.rateLimited,
            errorCode: 'rate_limit_error',
            retryAfter: 60
          };
        case 500:
        case 502:
        case 503:
        case 504:
          return {
            message: RESEARCH_CONFIG.errorMessages.serviceUnavailable,
            errorCode: 'service_unavailable',
            retryAfter: 30
          };
        default:
          return {
            message: RESEARCH_CONFIG.errorMessages.networkError,
            errorCode: 'api_error',
            retryAfter: 5
          };
      }
    }

    // Custom research errors
    if (error instanceof ResearchError) {
      return {
        message: error.message,
        errorCode: error.errorCode,
        retryAfter: error.retryAfter
      };
    }

    // Generic error
    return {
      message: RESEARCH_CONFIG.errorMessages.networkError,
      errorCode: 'unknown_error',
      retryAfter: 5
    };
  }

  /**
   * Get fallback response based on conversation stage
   */
  static getFallbackResponse(stage: string = 'general'): string {
    const fallbackKey = stage as keyof typeof RESEARCH_CONFIG.fallbackResponses;
    return RESEARCH_CONFIG.fallbackResponses[fallbackKey] || RESEARCH_CONFIG.fallbackResponses.general;
  }
}

/**
 * Retry wrapper for async functions
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  maxRetries: number = RESEARCH_CONFIG.maxRetries,
  delayMs: number = RESEARCH_CONFIG.retryDelayMs,
  exponentialBackoff: boolean = true
): Promise<T> {
  let lastError: Error;
  let currentDelay = delayMs;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error as Error;

      // Don't retry on validation errors, client errors, or CSP errors
      if (error instanceof ValidationError ||
          (error as any)?.status >= 400 && (error as any)?.status < 500 ||
          (error instanceof TypeError && error.message.includes('Content Security Policy'))) {
        throw error;
      }

      if (attempt === maxRetries) {
        throw lastError;
      }

      console.warn(`Attempt ${attempt + 1} failed, retrying in ${currentDelay}ms:`, error);

      await new Promise(resolve => setTimeout(resolve, currentDelay));

      if (exponentialBackoff) {
        currentDelay = Math.min(currentDelay * 2, 30000); // Max 30 seconds
      }
    }
  }

  throw lastError!;
}

/**
 * Timeout wrapper for async functions
 */
export async function withTimeout<T>(
  fn: () => Promise<T>,
  timeoutMs: number = RESEARCH_CONFIG.requestTimeoutMs
): Promise<T> {
  return Promise.race([
    fn(),
    new Promise<never>((_, reject) =>
      setTimeout(() => reject(new NetworkError('Request timed out')), timeoutMs)
    )
  ]);
}

/**
 * Safe execution wrapper that catches errors and returns fallback values
 */
export async function safeExecute<T>(
  fn: () => Promise<T>,
  fallbackValue: T,
  onError?: (error: Error) => void
): Promise<T> {
  try {
    return await fn();
  } catch (error) {
    console.error('Safe execution failed:', error);
    if (onError) {
      onError(error as Error);
    }
    return fallbackValue;
  }
}

/**
 * Debounce function for user input
 */
export function debounce<T extends (...args: any[]) => any>(
  func: T,
  waitMs: number
): (...args: Parameters<T>) => void {
  let timeoutId: NodeJS.Timeout;

  return (...args: Parameters<T>) => {
    clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), waitMs);
  };
}

/**
 * Check if error is retryable
 */
export function isRetryableError(error: any): boolean {
  // Don't retry validation errors or client errors
  if (error instanceof ValidationError) return false;
  if (error?.status >= 400 && error?.status < 500) return false;

  // Don't retry CSP errors
  if (error instanceof TypeError && error.message.includes('Content Security Policy')) return false;

  // Retry network errors, server errors, and timeouts
  return true;
}

/**
 * Format error for user display
 */
export function formatErrorForUser(error: any): string {
  const handled = ErrorHandler.handleApiError(error);
  return handled.message;
}

/**
 * Log error for debugging (only in development)
 */
export function logError(error: any, context?: string): void {
  if (process.env.NODE_ENV === 'development') {
    console.error(`Research Error${context ? ` (${context})` : ''}:`, error);
  }
}
